crates/core/src/db/message_log.rs

```diff
@@ -381,6 +397,8 @@ impl Iterator for IterSegment {
         if let Err(e) = self.read_exact_or_none(&mut buf)? {
             return Some(Err(e));
         }
+        self.read += HEADER_SIZE as u64;
```

Comment:
Should this add happen after the first read but before the second? Is failure in the middle visible?

Reply:
Hm right, this doesn't make much sense anymore.

For context: my original idea was to piggyback on this number to reset the log when replaying hits an un-decodable commit -- which almost (but only almost) certainly is the last message in the log. So in this case the byte offset should always fall on a message boundary.

I discarded the idea because it would require iteration to be synchronized with mutation of the log, which is undesirable if we are using those iterators for replication purposes. 

Now, appending to the log while reading concurrently is fine, but truncating is not. In practice, truncation would happen _before_ any readers could possibly obtain a handle (i.e. replaying, new leader term), but I'll need to noodle on it a bit more to find a way to statically ensure that's the case. Thoughts and ideas welcome!